test/envtest/k8s_objects_status_envtest_test.go

```diff
@@ -0,0 +1,198 @@
+//go:build envtest
+
+package envtest
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/samber/lo"
+	"github.com/samber/mo"
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8stypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	gatewayv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/controllers/gateway"
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/manager"
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/manager/featuregates"
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/util"
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/util/builder"
+)
+
+func TestHTTPRouteReconciliation_DoesNotBlockSyncLoopWhenStatusQueueBufferIsExceeded(t *testing.T) {
+	envcfg := Setup(t, scheme.Scheme)
+	ctrlClient := NewControllerClient(t, envcfg)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	gw := deployGateway(ctx, t, ctrlClient)
+	runManagerWithConfig(ctx, t, envcfg, gw, func(cfg *manager.Config) {
+		// Enable status updates and change the queue's buffer size to 0 to
+		// ensure that the status update notifications do not block the
+		// sync loop despite the fact that the status update queue is full.
+		// This is a regression test for https://github.com/Kong/kubernetes-ingress-controller/issues/4260.
+		// The test will timeout if the sync loop blocks (effectively manager.Run does not return after canceling context).
+		cfg.UpdateStatus = true
+		cfg.UpdateStatusQueueBufferSize = 0
+	})
+
+	backendService := corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: gw.Namespace,
+			Name:      "backend-svc",
+		},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{
+				{
+					Name:     "http",
+					Protocol: corev1.ProtocolTCP,
+					Port:     80,
+				},
+			},
+		},
+	}
+	require.NoError(t, ctrlClient.Create(ctx, &backendService))
+	t.Cleanup(func() { _ = ctrlClient.Delete(ctx, &backendService) })
+
+	httpRoute := gatewayv1beta1.HTTPRoute{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: gw.Namespace,
+			Name:      uuid.NewString(),
+		},
+		Spec: gatewayv1beta1.HTTPRouteSpec{
+			CommonRouteSpec: gatewayv1beta1.CommonRouteSpec{
+				ParentRefs: []gatewayv1beta1.ParentReference{{
+					Name: gatewayv1beta1.ObjectName(gw.Name),
+				}},
+			},
+			Rules: []gatewayv1beta1.HTTPRouteRule{{
+				Matches: []gatewayv1beta1.HTTPRouteMatch{{
+					Path: &gatewayv1beta1.HTTPPathMatch{
+						Type:  lo.ToPtr(gatewayv1beta1.PathMatchPathPrefix),
+						Value: lo.ToPtr("/path"),
+					},
+				}},
```

Comment:
You could also use builder for this: https://github.com/Kong/kubernetes-ingress-controller/blob/5f7eca1bbca498837e99def024307a6b762c9f75/internal/util/builder/httproutematch.go

Reply:
👍 eecb107